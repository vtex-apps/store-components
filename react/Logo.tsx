import React from 'react'
import { Link, useRuntime } from 'vtex.render-runtime'
import type { RenderContext } from 'vtex.render-runtime'
import { useCssHandles } from 'vtex.css-handles'
import type { CssHandlesTypes } from 'vtex.css-handles'
import * as Amp from 'react-amphtml'

import Placeholder from './components/Logo/LogoPlaceholder'
import styles from './components/Logo/styles.css'

const CSS_HANDLES = ['logoLink', 'logoImage', 'logoContainer'] as const

interface Props {
  /** Logo's URL */
  url: string
  /** Title to be displayed as alt text */
  title: string
  /** Logo's width */
  width?: string | number
  /** Logo's height */
  height?: string | number
  /** Link to redirect users on click */
  href?: string
  /** Logo's width on mobile */
  mobileWidth?: string | number
  /** Logo's height on mobile */
  mobileHeight?: string | number
  /** Used to override default CSS handles */
  classes?: CssHandlesTypes.CustomClasses<typeof CSS_HANDLES>
}

/**
 * Store's logo.
 */
function Logo({
  url,
  href,
  width,
  height,
  title,
  mobileWidth,
  mobileHeight,
  classes,
}: Props) {
  const {
    amp,
    account,
    hints: { mobile },
  } = useRuntime() as RenderContext

  const { handles } = useCssHandles(CSS_HANDLES, { classes })
  const logoClassNames = `store-logo ${handles.logoContainer} ${
    mobile ? styles.sizeMobile : styles.sizeDesktop
  }`

  const imgWidth = mobile && mobileWidth ? mobileWidth : width
  const imgHeight = mobile && mobileHeight ? mobileHeight : height

  const imageUrl = url?.replace(/{{account}}/g, account)

  let image = null

  if (amp && url) {
    image = (
      <Amp.AmpImg
        specName="default"
        width={imgWidth}
        height={imgHeight}
        alt={title}
        src={imageUrl}
        className={handles.logoImage}
      />
    )
  } else if (url) {
    image = (
      <img
        src={imageUrl}
        width={imgWidth}
        height={imgHeight}
        alt={title}
        className={handles.logoImage}
      />
    )
  }

  const logo = (
    <span className={`${logoClassNames} pv4 ph6`}>
      {url ? (
        image
      ) : (
        <Placeholder width={width} height={height} title={title} />
      )}
    </span>
  )

  return href ? (
    <Link to={href} className={handles.logoLink}>
      {logo}
    </Link>
  ) : (
    logo
  )
}

Logo.schema = {
  title: 'admin/editor.logo.title',
  description: 'admin/editor.logo.description',
  type: 'object',
  properties: {
    href: {
      title: 'admin/editor.logo.href.title',
      description: 'admin/editor.logo.href.description',
      type: 'string',
    },
    url: {
      title: 'admin/editor.logo.url.title',
      description: 'admin/editor.logo.url.description',
      type: 'string',
      widget: {
        'ui:widget': 'image-uploader',
      },
    },
    width: {
      title: 'admin/editor.logo.width.title',
      description: 'admin/editor.logo.width.description',
      type: 'number',
      isLayout: true,
    },
    height: {
      title: 'admin/editor.logo.height.title',
      description: 'admin/editor.logo.height.description',
      type: 'number',
      isLayout: true,
    },
    title: {
      title: 'admin/editor.logo.title.title',
      description: 'admin/editor.logo.title.description',
      type: 'string',
    },
    mobileWidth: {
      title: 'admin/editor.logo.mobileWidth.title',
      description: 'admin/editor.logo.mobileWidth.description',
      type: 'number',
      isLayout: true,
    },
    mobileHeight: {
      title: 'admin/editor.logo.mobileHeight.title',
      description: 'admin/editor.logo.mobileHeight.description',
      type: 'number',
      isLayout: true,
    },
  },
}

export default Logo
