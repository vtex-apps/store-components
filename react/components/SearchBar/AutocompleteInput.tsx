import React, { useRef, useEffect } from 'react'
import classNames from 'classnames'
import { Input, InputSearch } from 'vtex.styleguide'
import { ExtensionPoint, useChildBlock } from 'vtex.render-runtime'
import { IconSearch, IconClose } from 'vtex.store-icons'
import type { DownshiftProps } from 'downshift'
import { useIntl } from 'react-intl'

import { useSearchBarCssHandles } from './SearchBarCssHandles'

const DISPLAY_MODES = [
  'clear-button',
  'search-and-clear-buttons',
  'search-button',
]

/** Midleware component to adapt the styleguide/Input to be used by the Downshift */
export const CSS_HANDLES = [
  'autoCompleteOuterContainer',
  'compactMode',
  'externalSearchButtonWrapper',
  'paddingInput',
  'searchBarClearIcon',
  'searchBarIcon',
  'searchBarSearchIcon',
  'suffixWrapper',
] as const

const CloseIcon = () => {
  const hasIconBlock = Boolean(useChildBlock({ id: 'icon-close' }))

  if (hasIconBlock) {
    return <ExtensionPoint id="icon-close" size={16} type="line" />
  }

  return <IconClose type="line" size={16} />
}

const SearchIcon = () => {
  const hasIconBlock = Boolean(useChildBlock({ id: 'icon-search' }))

  if (hasIconBlock) {
    return <ExtensionPoint id="icon-search" />
  }

  return <IconSearch />
}

interface Props {
  /** Downshift prop to be passed to the input */
  autoComplete?: string
  /** Input ID */
  id?: string
  /** Downshift prop to be passed to the input */
  onBlur?: () => void
  /** Downshift prop to be passed to the input */
  onChange?: DownshiftProps<any>['onChange']
  /** Downshift prop to be passed to the input */
  onKeyDown?: (event: any) => void
  /** Downshift prop to be passed to the input */
  value?: string
  /** Downshift func to open the menu */
  openMenu?: () => void
  /** Placeholder to be used on the input */
  placeholder?: string
  compactMode?: boolean
  /** Clears the input */
  onClearInput: () => void
  /** Identify if the search icon is on left or right position */
  hasIconLeft?: boolean
  /** Custom classes for the search icon */
  iconClasses?: string
  /** Block class for the search icon */
  iconBlockClass?: string
  /** Identify if the search input should autofocus or not */
  autoFocus?: boolean
  /** Function to direct the user to the searchPage */
  onGoToSearchPage: () => void
  /**
   * @deprecated Use `displayMode`
   * Identify if icon should submit on click
   * */
  submitOnIconClick?: boolean
  /* Define the input display mode */
  displayMode?: 'clear-button' | 'search-and-clear-buttons' | 'search-button'
  /** Error message showed in search input */
  inputErrorMessage?: string
  /** The type of the search input */
  inputType?: 'text' | 'search'
  openAutocompleteOnFocus?: boolean
  onInputChange: DownshiftProps<any>['onChange']
  inputValue: string
}

function AutocompleteInput({
  onClearInput,
  compactMode,
  value,
  hasIconLeft,
  iconBlockClass,
  iconClasses,
  autoFocus,
  onGoToSearchPage,
  submitOnIconClick,
  displayMode = 'clear-button',
  openMenu,
  inputErrorMessage,
  inputType = 'text',
  openAutocompleteOnFocus,
  ...restProps
}: Props) {
  const { handles, withModifiers } = useSearchBarCssHandles()
  const inputRef = useRef<HTMLInputElement>(null)

  const intl = useIntl()

  let dMode = displayMode

  if (DISPLAY_MODES.indexOf(dMode) < 0) {
    console.error(
      `[store-components/search-bar] Invalid displayMode '${displayMode}'. The valid options are: ${DISPLAY_MODES.join(
        ', '
      )}`
    )
  }

  // for backward compatibility
  if (submitOnIconClick === true) {
    dMode = 'search-button'
  } else if (submitOnIconClick === false) {
    dMode = 'clear-button'
  }

  useEffect(() => {
    const changeClassInput = () => {
      // eslint-disable-next-line vtex/prefer-early-return
      if (compactMode && inputRef.current) {
        inputRef.current.placeholder = ''
        inputRef.current.classList.add(handles.paddingInput)
      }
    }

    changeClassInput()
    if (autoFocus) inputRef.current?.focus()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const hasValue = value != null && value.length > 0

  const showClearButton =
    (dMode === 'clear-button' && hasValue) ||
    dMode === 'search-and-clear-buttons'

  const showInternalSearchButton =
    (dMode === 'clear-button' && !hasValue) || dMode === 'search-button'

  const showExternalSearchButton = dMode === 'search-and-clear-buttons'

  const clearButton = showClearButton && (
    <button
      className={`${iconClasses ?? ''} ${withModifiers(
        'searchBarIcon',
        'clear'
      )} flex items-center pointer bn bg-transparent outline-0 pv0 pl0 pr3`}
      style={{
        visibility: hasValue ? 'visible' : 'hidden',
      }}
      aria-label={intl.formatMessage({
        id: 'store/search.clear-input',
      })}
      onClick={() => onClearInput()}
    >
      <CloseIcon />
    </button>
  )

  const internalSearchButton = showInternalSearchButton && (
    <button
      className={`${iconClasses ?? ''} ${withModifiers(
        'searchBarIcon',
        'search'
      )} flex items-center pointer bn bg-transparent outline-0 pv0 pl0 pr3`}
      onClick={() => hasValue && onGoToSearchPage()}
      aria-label={intl.formatMessage({
        id: 'store/search.submit-search',
      })}
    >
      <SearchIcon />
    </button>
  )

  const externalSearchButton = showExternalSearchButton && (
    <div
      className={`${handles.externalSearchButtonWrapper} bw1 bl b--muted-4 flex items-center `}
    >
      <button
        className={`${iconClasses ?? ''} ${withModifiers(
          'searchBarIcon',
          'external-search'
        )}  flex items-center h-100 pointer pv0 nr5 ph5 bn c-link`}
        onClick={onGoToSearchPage}
        aria-label={intl.formatMessage({
          id: 'store/search.submit-search',
        })}
      >
        <SearchIcon />
      </button>
    </div>
  )

  const suffix = (
    <div className={`${handles.suffixWrapper} flex h-100`}>
      {clearButton}
      {internalSearchButton}
      {externalSearchButton}
    </div>
  )

  const prefix = (
    <span
      className={`${iconClasses} ${withModifiers('searchBarIcon', 'prefix')} `}
    >
      <SearchIcon />
    </span>
  )

  const classContainer = classNames('w-100 flex', {
    [handles.compactMode]: compactMode,
  })

  if (inputType === 'search') {
    // <form> tag is needed for iOS:
    // https://stackoverflow.com/a/26287843
    return (
      <form
        action="#"
        onSubmit={e => {
          e.preventDefault()
          e.stopPropagation()
          e.nativeEvent.stopImmediatePropagation()
        }}
        className={handles.autoCompleteOuterContainer}
      >
        <div className={classContainer}>
          <InputSearch
            ref={inputRef}
            size="large"
            value={value}
            {...restProps}
            error={Boolean(inputErrorMessage)}
            errorMessage={inputErrorMessage}
            onSubmit={onGoToSearchPage}
            aria-label={intl.formatMessage({ id: 'store/search.autocompleteInput.aria-label' })}
          />
        </div>
      </form>
    )
  }

  return (
    <div className={handles.autoCompleteOuterContainer}>
      <div className={classContainer}>
        <Input
          ref={inputRef}
          size="large"
          value={value}
          prefix={hasIconLeft && prefix}
          suffix={suffix}
          {...restProps}
          error={Boolean(inputErrorMessage)}
          errorMessage={inputErrorMessage}
          aria-label={intl.formatMessage({ id: 'store/search.autocompleteInput.aria-label' })}
        />
      </div>
    </div>
  )
}

export default AutocompleteInput
