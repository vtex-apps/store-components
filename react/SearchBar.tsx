import React, { useState, useCallback, useEffect } from 'react'
import { useIntl } from 'react-intl'
import { ModalContext } from 'vtex.modal-layout'
import { useRuntime } from 'vtex.render-runtime'
import { useCssHandles } from 'vtex.css-handles'
import type { CssHandlesTypes } from 'vtex.css-handles'
import { usePixel } from 'vtex.pixel-manager'

import SearchBar, {
  CSS_HANDLES as SearchBarCssHandles,
} from './components/SearchBar/SearchBar'
import { CSS_HANDLES as AutocompleteInputCssHandles } from './components/SearchBar/AutocompleteInput'
import { SearchBarCssHandlesProvider } from './components/SearchBar/SearchBarCssHandles'

const { useModalDispatch } = ModalContext

export const SEARCH_BAR_CSS_HANDLES = [
  ...SearchBarCssHandles,
  ...AutocompleteInputCssHandles,
] as const

interface Props {
  /** Define when use the compact version of the component */
  compactMode?: boolean
  /** Define if the search icon is on left or right position */
  hasIconLeft?: boolean
  /** Custom classes for the search icon */
  iconClasses?: string
  /** Define if the search input should autofocus or not */
  autoFocus?: boolean
  /** Max width of the search bar */
  maxWidth?: string | number
  /**
   * Uses the term the user has inputed to try to navigate to the proper
   * page type (e.g. a department, a brand, a category)
   */
  attemptPageTypeSearch?: boolean
  /** A template for a custom url. It can have a substring ${term} used as placeholder to interpolate the searched term. (e.g. `/search?query=${term}`) */
  customSearchPageUrl?: string
  placeholder?: string
  /* Autocomplete Horizontal alignment */
  autocompleteAlignment?: 'right' | 'left' | 'center'
  /** Define if autocomplete should be open on input focus or not */
  openAutocompleteOnFocus?: boolean
  /** Define if input should blur on submit */
  blurOnSubmit?: boolean
  /** Define if icon should submit on click */
  submitOnIconClick?: boolean
  /** Minimum search term length allowed */
  minSearchTermLength?: number
  /** If true, the autocomplete will fill the whole window horizontally */
  autocompleteFullWidth?: boolean
  /** The type of the search input */
  inputType?: 'text' | 'search'
  /** Define the component display mode, such as which buttons should be visible */
  displayMode?: 'clear-button' | 'search-and-clear-buttons' | 'search-button'
  /** Define how the autocomplete component should be displayed. Possible values are: `overlay` (suggestions overlapping other components) and `container` (displays the suggestion within a container). */
  containerMode: 'overlay' | 'container'
  /** Used to override default CSS handles */
  classes?: CssHandlesTypes.CustomClasses<typeof SEARCH_BAR_CSS_HANDLES>
  /** The autocomplete can have touchable/clickable components. Interacting with those components may trigger blur and touch events that will close the autocomplete. When set to true, this prop will disable those handlers */
  disableBlurAndTouchEndHandler?: boolean
}

/**
 * Canonical search bar that uses the autocomplete endpoint to search for a specific product
 * */
function SearchBarContainer(props: Props) {
  const intl = useIntl()
  const {
    compactMode,
    hasIconLeft,
    iconClasses,
    autoFocus,
    maxWidth,
    attemptPageTypeSearch,
    customSearchPageUrl,
    placeholder = intl.formatMessage({
      id: 'store/search.placeholder',
    }),
    autocompleteAlignment = 'right',
    openAutocompleteOnFocus = false,
    containerMode = 'overlay',
    blurOnSubmit = false,
    submitOnIconClick,
    displayMode = 'clear-button',
    minSearchTermLength,
    autocompleteFullWidth = false,
    inputType = 'text',
    classes,
    disableBlurAndTouchEndHandler = false,
  } = props

  const modalDispatch = useModalDispatch()
  const { navigate, query } = useRuntime()
  const { push } = usePixel()

  const [inputValue, setInputValue] = useState(query?._q ?? '')

  useEffect(() => {
    setInputValue(query?._q ?? '')
  }, [query?._q])

  const { handles, withModifiers } = useCssHandles(SEARCH_BAR_CSS_HANDLES, {
    classes,
  })

  const closeModal = useCallback(() => {
    if (modalDispatch) {
      modalDispatch({
        type: 'CLOSE_MODAL',
      })
    }
  }, [modalDispatch])

  const handleClearInput = useCallback(() => {
    setInputValue('')
  }, [])

  const handleInputChange = useCallback(event => {
    setInputValue(event.target.value)
  }, [])

  const handleGoToSearchPage = useCallback(() => {
    const search = encodeURIComponent(inputValue.trim())

    push({
      event: 'search',
      term: search,
    })

    if (attemptPageTypeSearch) {
      window.location.href = `/${search}`
      closeModal()

      return
    }

    if (customSearchPageUrl) {
      navigate({
        to: customSearchPageUrl.replace(/\$\{term\}/g, search),
      })
      closeModal()

      return
    }

    // This param is only useful to track terms searched
    // See: https://support.google.com/analytics/answer/1012264
    const paramForSearchTracking = `&_q=${search}`

    navigate({
      page: 'store.search',
      params: { term: search },
      query: `map=ft${paramForSearchTracking}`,
      fallbackToWindowLocation: false,
    })
    closeModal()
  }, [
    inputValue,
    attemptPageTypeSearch,
    customSearchPageUrl,
    navigate,
    closeModal,
    push,
  ])

  return (
    <SearchBarCssHandlesProvider
      handles={handles}
      withModifiers={withModifiers}
    >
      <SearchBar
        // eslint-disable-next-line jsx-a11y/no-autofocus
        autoFocus={autoFocus}
        placeholder={placeholder}
        inputValue={inputValue}
        onClearInput={handleClearInput}
        onGoToSearchPage={handleGoToSearchPage}
        onInputChange={handleInputChange}
        compactMode={compactMode}
        hasIconLeft={hasIconLeft}
        iconClasses={iconClasses}
        maxWidth={maxWidth}
        attemptPageTypeSearch={attemptPageTypeSearch}
        customSearchPageUrl={customSearchPageUrl}
        autocompleteAlignment={autocompleteAlignment}
        openAutocompleteOnFocus={openAutocompleteOnFocus}
        blurOnSubmit={blurOnSubmit}
        submitOnIconClick={submitOnIconClick}
        displayMode={displayMode}
        minSearchTermLength={minSearchTermLength}
        autocompleteFullWidth={autocompleteFullWidth}
        inputType={inputType}
        containerMode={containerMode}
        disableBlurAndTouchEndHandler={disableBlurAndTouchEndHandler}
      />
    </SearchBarCssHandlesProvider>
  )
}

SearchBarContainer.schema = {
  title: 'admin/editor.search-bar.title',
}

export default SearchBarContainer
