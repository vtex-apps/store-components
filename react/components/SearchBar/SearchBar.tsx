import React, { useRef, useCallback, useState, useEffect, useMemo } from 'react'
import classNames from 'classnames'
import Downshift, {
  DownshiftProps,
  DownshiftState,
  StateChangeOptions,
} from 'downshift'
import debounce from 'debounce'
import {
  NoSSR,
  useRuntime,
  ExtensionPoint,
  useChildBlock,
} from 'vtex.render-runtime'
import { Overlay } from 'vtex.react-portal'
import { defineMessages, useIntl } from 'react-intl'
import { formatIOMessage } from 'vtex.native-types'

import styles from './SearchBar.css'
import AutocompleteResults from '../../AutocompleteResults'
import type { Props as AutocompleteResultsProps } from '../../AutocompleteResults'
import AutocompleteInput from './AutocompleteInput'
import { useSearchBarCssHandles } from './SearchBarCssHandles'

export const CSS_HANDLES = ['searchBarInnerContainer'] as const

const SEARCH_DELAY_TIME = 1000
const AUTOCOMPLETE_EXTENSION_ID = 'autocomplete-result-list'

const messages = defineMessages({
  searchTermTooShort: {
    id: 'store/search.search-term-too-short',
    defaultMessage: '',
  },
})

interface Props {
  /** Placeholder to be used on the input */
  placeholder: string
  /** Current value of the input */
  inputValue: string
  /** Function to handle input changes */
  onInputChange: DownshiftProps<any>['onChange']
  /** Function to direct the user to the searchPage */
  onGoToSearchPage: () => void
  /** Function to clear the input */
  onClearInput: () => void
  /** Indentify when use the compact version of the component */
  compactMode?: boolean
  /** Identify if the search icon is on left or right position */
  hasIconLeft?: boolean
  /** Custom classes for the search icon */
  iconClasses?: string
  /** Block class for the search icon */
  iconBlockClass?: string
  /** Identify if the search input should autofocus or not */
  autoFocus?: boolean
  /** Max width of the search bar */
  maxWidth?: string | number
  /** A template for a custom url. It can have a substring ${term} used as placeholder to interpolate the searched term. (e.g. `/search?query=${term}`) */
  customSearchPageUrl?: string
  /** Uses the term the user has inputed to try to navigate to the proper
   * page type (e.g. a department, a brand, a category)
   */
  attemptPageTypeSearch?: boolean
  /* Autocomplete Horizontal alignment */
  autocompleteAlignment?: 'right' | 'left' | 'center'
  /** Identify if autocomplete should be open on input focus or not */
  openAutocompleteOnFocus?: boolean
  /** Identify if input should blur on submit */
  blurOnSubmit?: boolean
  /** Identify if icon should submit on click */
  submitOnIconClick?: boolean
  /** Minimum search term length allowed */
  minSearchTermLength?: number
  /** If true, the autocomplete will fill the whole window horizontally */
  autocompleteFullWidth?: boolean
  /** The type of the search input */
  inputType?: 'text' | 'search'
  /** Define the component display mode,such as which buttons should be visible */
  displayMode?: 'clear-button' | 'search-and-clear-buttons' | 'search-button'
  /** Define how the autocomplete component should be displayed. Possible values are: `overlay` (suggestions overlapping other components) and `container` (displays the suggestion within a container). */
  containerMode: 'overlay' | 'container'
  /** The autocomplete can have touchable/clickable components. Interacting with those components may trigger blur and touch events that will close the autcomplete. When set to true, this prop will disable those handlers */
  disableBlurAndTouchEndHandler?: boolean
}

function SearchBar({
  placeholder,
  onInputChange,
  onGoToSearchPage,
  onClearInput,
  inputValue,
  compactMode,
  hasIconLeft,
  iconClasses,
  iconBlockClass,
  autoFocus,
  maxWidth,
  attemptPageTypeSearch,
  customSearchPageUrl,
  autocompleteAlignment,
  openAutocompleteOnFocus,
  blurOnSubmit,
  submitOnIconClick,
  displayMode,
  minSearchTermLength,
  autocompleteFullWidth,
  inputType,
  containerMode = 'overlay',
  disableBlurAndTouchEndHandler,
}: Props) {
  const { withModifiers } = useSearchBarCssHandles()
  const intl = useIntl()
  const container = useRef<HTMLDivElement>(null)
  const { navigate } = useRuntime()
  const [searchTerm, setSearchTerm] = useState(inputValue)
  const [inputErrorMessage, setInputErrorMessage] = useState<
    string | undefined
  >()

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const debouncedSetSearchTerm = useCallback(
    debounce((newValue: string) => {
      setSearchTerm(newValue)
    }, SEARCH_DELAY_TIME),
    []
  )

  useEffect(() => {
    debouncedSetSearchTerm(inputValue)
  }, [debouncedSetSearchTerm, inputValue])

  const onSelect = useCallback(
    element => {
      if (!element) {
        return
      }

      if (element.term) {
        if (attemptPageTypeSearch) {
          window.location.href = `/${element.term}`

          return
        }

        if (customSearchPageUrl) {
          navigate({
            to: customSearchPageUrl.replace(/\$\{term\}/g, element.term),
          })

          return
        }

        navigate({
          page: 'store.search',
          params: { term: element.term },
          query: 'map=ft',
        })

        return
      }

      let page = 'store.product'
      let params: Record<string, string> = {
        slug: element.slug,
        id: element.productId,
      }

      let query = ''
      const terms = element.slug.split('/')

      if (element.criteria) {
        // This param is only useful to track terms searched
        // See: https://support.google.com/analytics/answer/1012264
        const paramForSearchTracking = `&_c=${terms[0]}`

        page = 'store.search'
        params = { term: terms.join('/') }
        query = `map=c,ft${paramForSearchTracking}`
      }

      navigate({ page, params, query })
    },
    [navigate, attemptPageTypeSearch, customSearchPageUrl]
  )

  const validateInput = () => {
    if (minSearchTermLength && inputValue.length < minSearchTermLength) {
      return intl.formatMessage(messages.searchTermTooShort)
    }

    return null
  }

  const showInputErrorMessage = (newInputErrorMessage: string) => {
    setInputErrorMessage(newInputErrorMessage)
  }

  const hideInputErrorMessage = () => {
    setInputErrorMessage(undefined)
  }

  const fallback = (
    <AutocompleteInput
      placeholder={formatIOMessage({ id: placeholder, intl }) as string}
      onInputChange={onInputChange}
      inputValue={inputValue}
      hasIconLeft={hasIconLeft}
      iconClasses={iconClasses}
      iconBlockClass={iconBlockClass}
      inputErrorMessage={inputErrorMessage}
      onGoToSearchPage={onGoToSearchPage}
      onClearInput={onClearInput}
    />
  )

  const isAutocompleteDeclared = Boolean(
    useChildBlock({ id: AUTOCOMPLETE_EXTENSION_ID })
  )

  const SelectedAutocompleteResults = useMemo(() => {
    if (isAutocompleteDeclared) {
      const AutoCompleteResultsWrapper = (props: AutocompleteResultsProps) => (
        <ExtensionPoint id={AUTOCOMPLETE_EXTENSION_ID} {...props} />
      )

      return AutoCompleteResultsWrapper
    }

    const AutoCompleteResultsWrapper = (props: AutocompleteResultsProps) => (
      <AutocompleteResults {...props} />
    )

    return AutoCompleteResultsWrapper
  }, [isAutocompleteDeclared])

  const autocompleteContainerRef = useRef<HTMLDivElement>(null)

  const stateReducer = (
    state: DownshiftState<unknown>,
    changes: StateChangeOptions<unknown>
  ): StateChangeOptions<unknown> => {
    const { type } = changes

    if (
      (type === Downshift.stateChangeTypes.blurInput ||
        type === Downshift.stateChangeTypes.touchEnd) &&
      disableBlurAndTouchEndHandler
    ) {
      return {
        ...changes,
        isOpen: state.isOpen,
      }
    }

    return changes
  }

  return (
    <div
      ref={container}
      className={classNames('w-100 mw7 pv4', styles.searchBarContainer)}
      style={{
        ...(maxWidth && {
          maxWidth: typeof maxWidth === 'number' ? `${maxWidth}px` : maxWidth,
        }),
      }}
    >
      <NoSSR onSSR={fallback}>
        <Downshift onSelect={onSelect} stateReducer={stateReducer}>
          {({
            getInputProps,
            getItemProps,
            getMenuProps,
            selectedItem,
            highlightedIndex,
            isOpen,
            closeMenu,
            openMenu,
          }) => (
            <div
              className={`relative-m w-100 ${withModifiers(
                'searchBarInnerContainer',
                [isOpen ? 'opened' : '', inputValue ? 'filled' : '']
              )}`}
            >
              <AutocompleteInput
                // eslint-disable-next-line jsx-a11y/no-autofocus
                autoFocus={autoFocus}
                compactMode={compactMode}
                hasIconLeft={hasIconLeft}
                iconClasses={iconClasses}
                openAutocompleteOnFocus={openAutocompleteOnFocus}
                submitOnIconClick={submitOnIconClick}
                displayMode={displayMode}
                inputType={inputType}
                openMenu={openMenu}
                inputErrorMessage={inputErrorMessage}
                {...getInputProps({
                  onKeyDown: event => {
                    // Only call default search function if user doesn't
                    // have any item highlighted in the menu options
                    if (event.key === 'Enter' && highlightedIndex === null) {
                      const errorMessage = validateInput()

                      if (errorMessage) {
                        showInputErrorMessage(errorMessage)

                        return
                      }

                      if (blurOnSubmit) {
                        event.currentTarget.blur()
                      }

                      onGoToSearchPage()
                      closeMenu()
                    } else {
                      hideInputErrorMessage()
                    }
                  },
                  placeholder: formatIOMessage({
                    id: placeholder,
                    intl,
                  }) as string,
                  value: inputValue,
                  onChange: onInputChange,
                  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                  // @ts-expect-error this exists
                  onFocus: openAutocompleteOnFocus ? openMenu : undefined,
                })}
                onClearInput={onClearInput}
                onGoToSearchPage={onGoToSearchPage}
              />
              <div ref={autocompleteContainerRef} />

              {containerMode === 'container' ? (
                <SelectedAutocompleteResults
                  parentContainer={container ?? undefined}
                  {...{
                    attemptPageTypeSearch,
                    isOpen,
                    getMenuProps,
                    inputValue: searchTerm,
                    getItemProps,
                    selectedItem,
                    highlightedIndex,
                    closeMenu,
                    onClearInput,
                    customSearchPageUrl,
                  }}
                />
              ) : (
                <Overlay
                  fullWindow={autocompleteFullWidth}
                  alignment={autocompleteAlignment}
                  target={autocompleteContainerRef.current ?? undefined}
                >
                  <SelectedAutocompleteResults
                    parentContainer={container}
                    {...{
                      attemptPageTypeSearch,
                      isOpen,
                      getMenuProps,
                      inputValue: searchTerm,
                      getItemProps,
                      selectedItem,
                      highlightedIndex,
                      closeMenu,
                      onClearInput,
                      customSearchPageUrl,
                    }}
                  />
                </Overlay>
              )}
            </div>
          )}
        </Downshift>
      </NoSSR>
    </div>
  )
}

export default SearchBar
