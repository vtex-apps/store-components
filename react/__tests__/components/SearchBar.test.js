import React from 'react'
import { render, fireEvent, screen } from '@vtex/test-tools/react'

import SearchBar from '../../SearchBar'
import autocomplete from '../../graphql/autocomplete.gql'

const mockedUsePixelPush = jest.fn()

jest.mock('vtex.pixel-manager', () => ({
  usePixel: () => ({ push: mockedUsePixelPush }),
}))

describe('<SearchBar />', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  const mockedResult = {
    request: {
      query: autocomplete,
    },
    result: {
      data: {
        loading: false,
        autocomplete: {
          itemsReturned: [{ thumb: 'thumb', name: 'name', href: 'href' }],
        },
      },
    },
  }

  const renderComponent = (customProps = {}) => {
    return render(<SearchBar {...customProps} />, {
      graphql: { mocks: [mockedResult] },
    })
  }

  it('should be able to mount and not break', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toBeTruthy()
  })

  it('should match snapshot', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })

  it('should have CSS handle searchBarIcon', () => {
    const { container } = renderComponent()
    const element = container.querySelector('.searchBarIcon')

    expect(element).toBeTruthy()
  })

  it('should display internal search icon button', () => {
    const { container } = renderComponent({
      displayMode: 'search-button',
    })

    expect(container.querySelector('.searchBarIcon--search')).toBeTruthy()
  })

  it('should display internal clear icon button when a value is typed', () => {
    const { container } = renderComponent({
      displayMode: 'search-button',
      value: 'foo',
    })

    expect(container.querySelector('.searchBarIcon--search')).toBeTruthy()
  })

  it('should display external search icon button', () => {
    const { container } = renderComponent({
      displayMode: 'search-and-clear-buttons',
    })

    expect(
      container.querySelector('.searchBarIcon--external-search')
    ).toBeTruthy()
  })

  it('should display interal clear icon button when a value is typed', () => {
    const { container } = renderComponent({
      displayMode: 'search-and-clear-buttons',
      value: 'foo',
    })

    expect(container.querySelector('.searchBarIcon--clear')).toBeTruthy()
  })

  it('should trigger vtex:search when a value is typed', async () => {
    const { container } = renderComponent({
      displayMode: 'search-and-clear-buttons',
      submitOnIconClick: true,
    })

    const [searchButton] = container.getElementsByClassName('searchBarIcon')
    const inputNode = screen.getByPlaceholderText('Search')

    fireEvent.change(inputNode, { target: { value: 'foo' } })

    fireEvent.click(searchButton)

    const expectedPixelEvent = {
      event: 'search',
      term: 'foo',
    }

    expect(mockedUsePixelPush).toHaveBeenCalledWith(expectedPixelEvent)
  })
})
