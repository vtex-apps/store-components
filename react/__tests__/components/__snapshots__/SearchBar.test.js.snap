// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-100 mw7 pv4 searchBarContainer"
  >
    <div
      class="NoSSR-mock"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="downshift-1-label"
        class="relative-m w-100 searchBarInnerContainer"
        role="combobox"
      >
        <div
          class="autoCompleteOuterContainer"
        >
          <div
            class="w-100 flex"
          >
            <label>
              <input
                aria-autocomplete="list"
                aria-labelledby="downshift-1-label"
                autocomplete="off"
                data-error="false"
                id="downshift-1-input"
                placeholder="Search"
                value=""
              />
              <div
                class="suffixWrapper flex h-100"
              >
                <button
                  aria-label="Search Products"
                  class=" searchBarIcon searchBarIcon--search flex items-center pointer bn bg-transparent outline-0 pv0 pl0 pr3"
                >
                  <div
                    class="extension-point-icon-search"
                  />
                </button>
              </div>
            </label>
          </div>
        </div>
        <div />
        <div>
          <div
            class="extension-point-autocomplete-result-list"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
